# Assignment  3

2018136073 위 건

### [일 반 문 제]

### 1. L = [1, 2, 3, 4, 5] 일 때, 다음 각 문장을 수행한 후의 결과를 보고 납득할 만한 이유를 설명하시오.

>- 1)  
    >    L[1:3] = [100]  
    >    print(L)

In [4]:
L = [1, 2, 3, 4, 5]
L[1:3] = [100]
print(L)

[1, 100, 4, 5]


#### < 결과 >  
 리스트 L의 인덱스는 0 ~ 4까지 존재하며,  
 2번 줄의 명령을 통해 L리스트의 L[1:3] 즉, 1번 인덱스부터 3번 인덱스 미만의 값이 [2, 3]이 아닌 [100]을 가리키게 되었기에  
 그 자리를 대체하여 들어가고 결과적으로 [1, 100, 4, 5]가 출력된다.

>- 2)  
    >    L[1:3] = [(100)]  
    >    print(L)

In [13]:
L = [1, 2, 3, 4, 5]
L[1:3] = [(100)]
print(L)

[1, 100, 4, 5]


#### < 결과 >  
 리스트 L의 인덱스는 0 ~ 4까지 존재하며,  
 2번 줄의 명령을 통해 L리스트의 L[1:3] 즉, 1번 인덱스부터 3번 인덱스 미만의 값이 [2, 3]이 아닌 [(100)]을 가리키게 되었기에  
 그 자리를 대체하여 들어가고 결과적으로 [1, 100, 4, 5]가 출력된다.  
 또한, (100)의 값은 ","를 사용하지 않았기에 튜플이 아닌 정수값으로 들어가게 된다.

>- 3)  
    >    L[1:3] = 100  
    >    print(L)

In [16]:
L = [1, 2, 3, 4, 5]
L[1:3] = 100
print(L)

TypeError: can only assign an iterable

#### < 결과 >  
 리스트 L의 슬라이싱 L[1:3]은 interable 값만 가리킬 수 있다.  
 즉, Sequence 자료형이 아닌 정수 자료형인 100은 오류가 발생한다.  

>- 4)  
    >    L[1:3] = (100)  
    >    print(L)

In [1]:
L = [1, 2, 3, 4, 5]
L[1:3] = (100)
print(L)

TypeError: can only assign an iterable

#### < 결과 >  
 4번과 마찬가지로, 리스트 L의 슬라이싱 L[1:3]은 interable 값만 가리킬 수 있다.  
 즉, Sequence 자료형이 아닌 정수 자료형인 (100)은 오류가 발생한다.  
 (100,)이 아닌 (100)은 정수 자료형이다.

>- 5)  
    >    L[1:3] = (100, 101, 102)  
    >    print(L)

In [23]:
L = [1, 2, 3, 4, 5]
L[1:3] = (100, 101, 102)
print(L)

<class 'list'>
[1, 100, 101, 102, 4, 5]


#### < 결과 >  
 리스트 L의 인덱스는 0 ~ 4까지 존재하며,  
 2번 줄의 명령을 통해 L리스트의 L[1:3] 즉, 1번 인덱스부터 3번 인덱스 미만의 값이 [2, 3]이 아닌 (100, 101, 102)를  
 가리키게 되었기에 그 자리를 대체하여 들어가고 결과적으로 [1, 100, 101, 102, 4, 5]가 출력된다.  
 또한 시퀀스 자료형 끼리의 대입에서 값을 추가할 때에는 각 인덱스에 맞게 한번에 하나의 요소를 반환하여 가리키게한다.  
 즉, 각 인덱스에 해당 요소들이 하나씩 추가된다.  

>- 6)  
    >    L[1:3] = [100, 101, 102]  
    >    print(L)

In [24]:
L = [1, 2, 3, 4, 5]
L[1:3] = [100, 101, 102]
print(L)

[1, 100, 101, 102, 4, 5]


#### < 결과 >  
 리스트 L의 인덱스는 0 ~ 4까지 존재하며,  
 2번 줄의 명령을 통해 L리스트의 L[1:3] 즉, 1번 인덱스부터 3번 인덱스 미만의 값이 [2, 3]이 아닌 [100, 101, 102]를  
 가리키게 되었기에 그 자리를 대체하여 들어가고 결과적으로 [1, 100, 101, 102, 4, 5]가 출력된다.  
 또한, [100, 101, 102]의 리스트는 각 요소를 하나씩 반환하여 L 리스트가 가리키게 되므로, 차례로 1번 인덱스부터 100, 101, 102를 가리킨다.

>- 7)  
    >    L[1:3] = [(100, 101, 102)]  
    >    print(L)

In [25]:
L = [1, 2, 3, 4, 5]
L[1:3] = [(100, 101, 102)]
print(L)

[1, (100, 101, 102), 4, 5]


#### < 결과 >  
 리스트 L의 인덱스는 0 ~ 4까지 존재하며,  
 2번 줄의 명령을 통해 L리스트의 L[1:3] 즉, 1번 인덱스부터 3번 인덱스 미만의 값이 [2, 3]이 아닌 [(100, 101, 102)]를  
 가리키게 되었기에 그 자리를 대체하여 들어가고 결과적으로 [1, (100, 101, 102), 4, 5]가 출력된다.  
 다만, 이번에는 5번 문제와 다르게 리스트 내부에 튜플이 포함된 형태이기에 리스트 [(100, 101, 102)]의 요소는 튜플 (100, 101, 102) 밖에 존재하지 않는다. 그렇기에 들어가는 값 또한 100, 101, 102가 아닌 (100, 101, 102) 1개다.

>- 8)  
    >    L[1:3] = [[100, 101, 102]]  
    >    print(L)

In [26]:
L = [1, 2, 3, 4, 5]
L[1:3] = [[100, 101, 102]]
print(L)

[1, [100, 101, 102], 4, 5]


#### < 결과 >  
 리스트 L의 인덱스는 0 ~ 4까지 존재하며,  
 2번 줄의 명령을 통해 L리스트의 L[1:3] 즉, 1번 인덱스부터 3번 인덱스 미만의 값이 [2, 3]이 아닌 [[100, 101, 102]]를  
 가리키게 되었기에 그 자리를 대체하여 들어가고 결과적으로 [1, [100, 101, 102], 4, 5]가 출력된다.  
 7번 문제와 같이 리스트 내부에 리스트가 포함된 형태이기에 리스트 [[100, 101, 102]]의 요소는 리스트 [100, 101, 102] 밖에 존재하지 않는다. 그렇기에 들어가는 값 또한 100, 101, 102가 아닌 [100, 101, 102] 1개다.

### 2. 문자열 S = "Hello World and Python"에 대해 다음 요구사항에 대한 Python 코드를 제시하시오.  
>- 1) 단어의 순서가 역순으로 되어진 문자열("Python and World Hello")을 만들고  
>- 2) 1)의 결과 문자열에 대해 메소드 split과 join을 이용하여 공백을 모두 없앤 문자열 "PythonandWorldHello"을 만드시오.  

In [30]:
S = "Hello World and Python"
revS = " ".join(S.split()[::-1])
print("".join(revS.split()))

PythonandWorldHello


#### < 결과 >  
 S = "Hello World and Python"이라는 문자열을 갖고,
 2번 줄에서 revS라는 역순으로 되어진 문자열을 가리키게 한다.  
 S에 split()을 활용하여 공백을 기준으로 한 리스트 ['Hello', 'World', 'and', 'Python']을 만든다.  
 그 다음, " ".join을 통하여 리스트의 요소들을 " "을 띄워서 문자열로 다시 뭉친다. ==> ["Python", 'and', 'World', 'Hello']  
 완료되었다면 3번줄에서 revS를 다시 split()하여 공백을 기준으로 자른 다음 리스트에 넣고 ["Python", 'and', 'World', 'Hello']을 만든다.  
 그 후, "".join()함수를 통해 해당 리스트를 공백없는 문자열로 만든다음 출력한다.

### 3. 다음 코드를 보고 물음에 답하시오.
> a = [1, 2,3]  
> b = a * 3  
> c = [a] * 3  
> print(b)  
> print(c)  
>- 1) 위 코드를 수행한 이후 다음 코드를 수행한 후 b와 c의 내용을 확인하시오.  
    >          a[0] = 0  
>- 2) b와 c객체의 차이는 무엇인가?  

In [5]:
a = [1, 2,3]
b = a * 3
c = [a] * 3
print("   < a[0] = 0 실행 전 >")
print(b)
print(c)
#1
a[0] = 0
print("\n   < a[0] = 0 실행 후 >")
print(b)
print(c)

   < a[0] = 0 실행 전 >
[1, 2, 3, 1, 2, 3, 1, 2, 3]
[[1, 2, 3], [1, 2, 3], [1, 2, 3]]

   < a[0] = 0 실행 후 >
[1, 2, 3, 1, 2, 3, 1, 2, 3]
[[0, 2, 3], [0, 2, 3], [0, 2, 3]]


#### <결과>
> 1) 위 코드를 수행한 이후 다음 코드를 수행한 후 b와 c의 내용을 확인하시오.  
   >- b = [1, 2, 3, 1, 2, 3, 1, 2, 3]  
   >- c = [ [0, 2, 3], [0, 2, 3], [0, 2, 3] ]  
    
> 2) b와 c객체의 차이는 무엇인가?  
>> b 객체의 경우 a객체가 바라보는 1, 2, 3의 각 주소값을 순서대로 3번 연속으로 바라보고,  
>> c 객체의 경우 모든 인덱스가 a객체를 바라본다.  
>> 이러한 상황에서 a객체의 0번 인덱스가 1이 아닌 0을 바라본다면,  
>> b 객체의 경우 그대로 1, 2, 3의 주소값을 바라보기에 변화가 없지만,  
>> c 객체의 경우 a객체를 바라보았기에, a 객체의 0번 인덱스가 1에서 0으로 바뀐 것처럼 c 객체에서도 1이 아닌 0으로 시작한다.  

### 4. 다음 문자열을 ' : ' 기준으로 분리하여 리스트로 만들고 각 문자열의 좌우 공백을 제거하시오 ( S -> I)
> s = '&nbsp;&nbsp;first star&nbsp;&nbsp;&nbsp;:&nbsp;&nbsp;&nbsp;second star&nbsp;&nbsp;&nbsp;:&nbsp;&nbsp;&nbsp;&nbsp;third star&nbsp;&nbsp;'  
> I = ['first star', 'second star', 'third star']   
>  
>- 1) for ~ in 문을 이용한 코드를 작성하시오.  
>- 2) 리스트 내포를 이용한 코드를 작성하시오.

In [8]:
s = '  first star   :   second star   :    third star  '
mid = s.split(':')
l = [i.strip() for i in mid]
print("s = " + s)
print("l = {0}".format(l))

s =   first star   :   second star   :    third star  
l = ['first star', 'second star', 'third star']


#### <결과>
> s 문자열에서 split을 통해 ':'을 기준으로 나눈 문자열을 mid라는 list로 변환한 뒤,  
> for i in mid를 통하여 mid안의 각 인덱스 값들을 순서대로 i가 가리키게 한 뒤, 그 값을 .strip()으로 양 공백을 제거한다.  
>- 1) for ~ in 문을 이용한 코드를 작성하시오.  
>> for i in mid  => mid 안에 있는 각 요소들을 i가 가리키게 한다.
>- 2) 리스트 내포를 이용한 코드를 작성하시오.  
>> l = [i.strip() for i in mid] => 반복을 통해 생성되는 i의 양 공백을 제거한 뒤 l 리스트에 삽입한다.

### 5. 다음과 같이 0보다 큰 정수 리스트 변수 list를 인자로 받는 함수 addall(list)와 addallodd(list)를 for ~ in 리터럴과 리스트 내포 방식으로 각각 제시하시오.
>- 1) addall(list) 함수 (리스트 내의 모든 정수를 더하는 함수로서 해답으로는 for ~ in 리터럴과 리스트 내포 방식으로 각각 제시하시오.)
>> - addall([1])  
>> 1  
>> - addall([1, 2, 3, 4, 5, 6, 7, 8, 9])  
>> 45  
>- 2) addallodd(list) 함수 (리스트 내의 모든 홀수를 더하는 함수로서 해답으로는 for ~ in 리터럴과 리스트 내포 방식으로 각각 제시하시오.)  
>> - addallodd([1])  
>> 1  
>> - addallodd([1, 2, 3, 4, 5, 6, 7, 8, 9])  
>> 25  

>- [참고 1]:리스트 내포 방식으로 addall() 및 addalldoo()함수를 만들 때엔 리스트 내포에 의해 생성된 리스트 결과에 대해 내장함수를 사용하는 방식을 고려해보시오. (주의 : 리스트 내포 방식의 addall() 및 addallodd() 함수의 몸체는 단 1줄로 작성되어야 한다.)  
>- [참고 2]: sum() 등의 내장함수 사용 가능함  

In [2]:
# for ~ in 리터럴
## addall()
def addall(list):
    sum = 0
    for i in list:
        if i > 0: sum += i
    return sum

## addallodd()
def addallodd(list):
    sum = 0
    for i in list:
        if i > 0 and i % 2 : sum += i
    return sum
list1 = [1]
listMany = [1, 2, 3, 4, 5, 6, 7, 8, 9]

print("1) for ~ in 리터럴")
print("addall() 함수")
print(addall(list1))
print(addall(listMany))
print("addallodd() 함수")
print(addallodd(list1))
print(addallodd(listMany))

# 리스트 내포
## addall()
def addall(list):
    return sum([i for i in list if i > 0])
## addallodd()
def addallodd(list):
    return sum([i for i in list if i > 0 and i % 2])

print("\n2) 리스트 내포")
print("addall() 함수")
print(addall(list1))
print(addall(listMany))
print("addallodd() 함수")
print(addallodd(list1))
print(addallodd(listMany))

1) for ~ in 리터럴
addall() 함수
1
45
addallodd() 함수
1
25

2) 리스트 내포
addall() 함수
1
45
addallodd() 함수
1
25


#### <결과>
>- 1) for ~ in 리터럴 함수    
>>- addall(list) 함수  
>> 합을 저장할 변수 sum을 0으로 초기화 한 후,  
>> for i in list:를 통하여 i에 list 요소들을 순서대로 대입한다.  
>> 그 후, if i > 0이라면 sum에 i를 추가한다.  
>>  &nbsp;
>>- addallodd(list) 함수  
>> addall(list) 함수에서 조건에 i % 2 즉, 2로 나누어 떨어지지 않을 때를 추가한다.

>- 2) 리스트 내포 함수  
>>- addall(list) 함수  
>> result = [i for i in list if i > 0]  
>> list 안의 요소를 i에 대입 후 i가 0보다 크다면 result 안에 i를 삽입한다.  
>> 그 후, return sum(result)를 통하여 result 리스트 안의 모든 요소를 더한 값을 반환한다.  
>>  &nbsp;
>>- addallodd(list) 함수  
>> addall(list) 함수에서 조건에 if i % 2를 추가하여 홀수일 때의 i를 삽입한다.  


### 6. 다음 코드를 보고 물음에 답하시오.
> L1 = [1, 2, 3]  
> L2 = [4, 5, 6]  
> d = {'low':L1, 'high':L2}  
> e = d  
> f = d.copy()  
> print(d)  
> print(e)  
> print(f)  
> print()  
> d['low'] = [10, 20, 30]  
> d['high'][1] = 500  
> print(d)  
> print(e)  
> print(f)  
>- 1) 위 코드의 수행 결과 d와 e는 항상 같은 결과를 출력한다. 그 이유는 무엇인가?  
>- 2) 위 코드의 수행 결과 마지막 f의 내용은 다음과 같다. 이 때 'high' 키의 값 내용 중 500이 출력된 이유와 'low'키의 값 내용 중 [1, 2, 3]이 출력된 이유를 설명하시오 ([hint] : shallow copy 관점에서 설명하시오.)  
>  {'high' : [4, 500, 6], 'low', [1, 2, 3]}

In [52]:
L1 = [1, 2, 3]
L2 = [4, 5, 6]
d = {'low':L1, 'high':L2}
e = d
f = d.copy()
print(d)
print(e)
print(f)
print()
d['low'] = [10, 20, 30]
d['high'][1] = 500
print(d)
print(e)
print(f)
print()

print(id(d['low']))
print(id(L1))
print(id(f['high']))
print(id(L2))


{'low': [1, 2, 3], 'high': [4, 5, 6]}
{'low': [1, 2, 3], 'high': [4, 5, 6]}
{'low': [1, 2, 3], 'high': [4, 5, 6]}

{'low': [10, 20, 30], 'high': [4, 500, 6]}
{'low': [10, 20, 30], 'high': [4, 500, 6]}
{'low': [1, 2, 3], 'high': [4, 500, 6]}

1980737436608
1980735556864
1980736333248
1980736333248


##### <풀이> 
>- 1) d와 e는 얕은 복사로 같은 주소값을 참조하기 때문인다.  
>- 2 - 1) 'high' 키의 경우 d['high'][1] = 500을 통해 d['high'] 즉, L2의 주소에서 [1]째의 값을 직접 500으로 바꾸었기에,  
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; L2를 가리키는 f의 'high' 또한 [4, 500, 6]으로 바뀌었다.
>- 2 - 2) 'low' 키의 경우 d['low']가 가리키는 리스트의 주소가 [1, 20, 30]의 1980737436608로 바뀌었지만, f의 경우는 그대로 L1의  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1980735556864을 가리키고 있기에 변하지 않는다.  
* 즉 d와 e의 경우같은 얕은 복사는 같은 대상의 주소를 참조하기 때문에, 한쪽에 변화가 있을 시, 다른 쪽도 똑같이 변하지만, d와 f와 같은 깊은 복사의 경우는 새로운 객체를 만들어서 참조하고, 내부 요소를 직접적으로 바꾸지 않는 한 변화가 없다.

### 7. 사전 d = {'one' : 1, 'two' : 2, 'three' : 3, 'four' : 4, 'five' : 5}가 주어졌을 때 다음 요구사항에 맞는 코드를 제시하시오}
>- [참고] : d.keys(), d.values()를 통해 리스트를 얻어낸 후 리스트가 지니고 있는 sort(cmp = None, key = None, reverse = False)함수를 활용하시오.  
>- 1) 키의 알파벳 오름차순 순서대로 튜플(키, 값)을 차례대로 출력하시오.  
>- 2) 키의 알파벳 내림차순 순서대로 튜플 (키, 값)을 차례대로 출력하시오.  
>- 3) 값의 오름차순 순서대로 튜플 (키, 값)을 차례대로 출력하시오.  
>- 4) 값의 내림차순 순서대로 튜플 (키, 값)을 차례대로 출력하시오.  

In [107]:
d = {'one':1, 'two':2, 'three':3, 'four':4, 'five':5}
a = list(zip(d.keys(),d.values()))

# 1)
a.sort(key = lambda x:x[0])
print("1) 알파벳 오름차순 : ", end=" ")
for i in a:
    print(i, end=" ")
print()

# 2)
a.sort(key = lambda x:x[0],reverse = True)
print("2) 알파벳 내림차순 : ", end=" ")
for i in a:
    print(i, end=" ")
print()

# 3)
a.sort(key = lambda x:x[1])
print("3) 값 오름차순 : ", end=" ")
for i in a:
    print(i, end=" ")
print()

# 4)
a.sort(key = lambda x:x[1],reverse = True)
print("4) 값 내림차순 : ", end=" ")
for i in a:
    print(i, end=" ")

1) 알파벳 오름차순 :  ('five', 5) ('four', 4) ('one', 1) ('three', 3) ('two', 2) 
2) 알파벳 내림차순 :  ('two', 2) ('three', 3) ('one', 1) ('four', 4) ('five', 5) 
3) 값 오름차순 :  ('one', 1) ('two', 2) ('three', 3) ('four', 4) ('five', 5) 
4) 값 내림차순 :  ('five', 5) ('four', 4) ('three', 3) ('two', 2) ('one', 1) 

##### <풀이> 
>- zip함수를 통해 (d.keys, d.values)를 튜플로 묶고, list화 시켜 a에 대입한다.  
>- 1) a.sort()를 활용하여 기본적으로 오름차순 정렬을 해주는데, lambda x:x[0]을 요소로 들어오는 튜플의 첫번째 항(키 값)을 기준으로 한다.
>- 2) 1)번 내용에서 sort()함수의 기본적으로 내포된 reverse = False 값을 True로 변환시켜 내림차순 정렬을 해준다.
>- 3) a.sort()를 활용하여 기본적으로 오름차순 정렬을 해주는데, lambda x:x[1]을 요소로 들어오는 튜플의 두번째 항(값)을 기준으로 한다.
>- 4) 3)번 내용에서 sort()함수의 기본적으로 내포된 reverse = False 값을 True로 변환시켜 내림차순 정렬을 해준다.

### 8. 다음 문제를 정확히 읽고 적합한 코딩을 제시하시오.
>- **length(X)** 를 다음과 같이 정의한다 : 10진수로 표현된 X의 자릿수  
>- **Strange**라는 숫자를 아래와 같이 정의한다.  
>>- 음이 아닌 정수 중 자릿수가 1인 모든 숫자는 **Strange**이다.
>>- **length(X)** > 1인 숫자 X는 다음 2가지를 만족하면 **Strange**이다.
>>>- X는 **length(X)** 로 정확히 나누어 떨어진다.
>>>- X/length(X)가 **Strange**이다.

> 이 문제에서 숫자하나가 주어질 때, 이 숫자가 Strange인지 판단하는 함수 isStrange(num)을 작성하시오.  
> isStrange(num) 함수가 재귀적인 함수가 되도록 작성해보세요.  

In [164]:
# 10진수로 표현된 X의 자릿수
def length(num):
    return len(str(num))

# 주어진 숫자가 Strange인지 판단하는 함수
def isStrange(num):
    if num >= 0:
        if length(num) == 1:
            return True
        elif length(num) > 1 and num % length(num) == 0:
            return isStrange(num // length(num))    
    return False

print(isStrange(0))
print(isStrange(56))
print(isStrange(100))
print(isStrange(7882602506492313600))
print(isStrange(6012302549522098651))

True
True
False
True
False


##### <풀이> 
>- def length(num) 함수는 num을 문자열로 받은 뒤, 그에 대한 길이를 구하는 함수로 자릿수를 10의 진수로 반환한다.
>- def isStrange(num) 함수는 입력 받은 숫자 num이 0을 포함한 양의 정수일 때,
>> 자릿수가 1인 수라면 True  
>> 또, 2 이상의 자릿수를 가지며, length(num)으로 나누어 떨어지는 수일 때, isStrange(num//length(num))을 통해 몫을 정수로 받는 재귀함수를 활용하여 1의 자릿수까지 간다면 True, 도중에 조건에 맞지 않는다면 False를 반환한다.  

> 음수, 조건이 안맞으면 False를 반환한다.  

### 9. 다음 문제를 정확히 읽고 적합한 코딩을 제시하시오.
>- 1번 그룹, 2번 그룹, ... n번 그룹으로 번호가 매겨진 **총 n개의 그룹**이 존재하고, i<sup>th</sup> 그룹에는 임의의 팀원 수 **a**<sub>i</sub> 명이 있다.  
하나의 버스 정류장이 있고, 버스는 오직 한 대만 운영한다. 버스는 버스 정류장에 도착하고 모든 그룹은 버스를 통해 여행하길 원한다.  
하지만, 같은 그룹에 속한 팀원들은 분리되기를 원치 않는다. 그래서 그룹의 모든 팀원이 탈 수 있을 때 에만 그 그룹은 버스에 탑승한다.    
또한, 이들은 그룹의 번호순으로 여행을 하길 원한다.  
예를 들어, 3번 그룹은 1번 그룹과 2번 그룹이 여행을 한 뒤에 버스를 탈 수 있거나, 1번 그룹과 2번 그룹이 이미 버스에 탑승한 상태에 있을 때 동시에 버스를 탈 수 있다.  
마지막으로, 버스는 빈 좌석이 없을 때에만 여행을 떠난다.  

>- 이 문제에서 각 그룹의 팀원 수가 리스트로 주어질 때, 가능한 모든 버스의 사이즈 **x**(x명의 사람들이 동시에 탈 수 있음)를 반환하는 함수 solve(a) 를 작성하시오.

>- [주의] 반드시 계산 복잡도를 최대한 낮추세요. 아래 예시 3번을 실행했을 때 결과가 바로 나타날 수 있도록!!

In [10]:
import time

# 약수를 구하는 함수
def getDivisor(n):
    d = []
    for i in range(1, int(n**(1/2)) + 1):
        if (n % i == 0):
            d.append(i) 
            if ( (i**2) != n) : 
                d.append(n // i)
    d.sort()
    return d


# 버스의 사이즈를 반환하는 함수
def solve(a):
    result = []
    sumL = 0
    copyList = []
    max = list(set(a))[-1]
    div = getDivisor(sum(a))
    for i in div:
        if i >= max:
            copyList.append(i)
    for i in copyList:
        count = 1
        for j in a:
            sumL += j
            if sumL == i:
                sumL = 0
            elif sumL > i:
                count = 0
                break
        if count == 1:
            result.append(i)
    return result

a = [1, 2, 1, 1, 1, 2, 1, 3]
b = [1, 1, 1, 1]
c = [9999]*10**5
print(solve(a))
print(solve(b))
start = time.time()
print(solve(c))
end = time.time()
print('소요된 시간 : ', end - start)

[3, 4, 6, 12]
[1, 2, 4]
[9999, 1999800, 2499750, 3999600, 4999500, 6249375, 7999200, 9999000, 12498750, 19998000, 24997500, 31246875, 39996000, 49995000, 62493750, 99990000, 124987500, 199980000, 249975000, 499950000, 999900000]
소요된 시간 :  0.16757941246032715


##### <풀이> 
>- 1) 약수를 구하는 함수
>     임의의 자연수 N의 약수들 중에서 두 약수의 곱이 N이 되는 약수 A와 약수 B는 반드시 존재하기 때문에 그 공식을 활용하여 함수를 작성한다.
>- 2) 버스의 사이즈를 반환하는 함수
>     버스의 사이즈의 최대값은 결국 주어지는 리스트의 모든 값을 더한 값이다. (모두 여행을 떠나는 상황)
>     그리고 버스의 사이즈로 가능한 수들은 그 최대값의 약수이며, 그 중, 주어진 리스트의 요소 중 가장 큰 값보다 크거나 같은 수부터 시작하여야 한다.  

>     약수가 포함된 div리스트 안의 요소를 i에 대입하여 반복하는 반복문  
    >>    div 리스트의 약수 중, 그룹 리스트의 가장 큰 수보다 크거나 같은 수들을 newList에 삽입한다.  
     
>     copyList 안의 요소를 i에 대입하여 반복하는 반복문
    >>    앞 그룹부터 출발하여 copyList안의 숫자로 정확하게 다 나누어지는 지 판단하여 되면 result 리스트에 담고 아니면 pass한다. 

### [Incremental Project 문제]

이전 Assignment 2의 마지막 문제는 웹 URL로 지정된 웹페이지를 문자열로 가져와 모든 HTML 태그를 제외한 순수 텍스트를 얻어내고 그 안에 존재하는 단어를 추출하고 단어들의 총 개수를 출력하는 프로그램을 작성하는 것이었다. 이번에는 그 마지막 숙제를 그대로 확장하여 웹 URL로 지정된 웹페이지 내 순수 텍스트 안에 존재하는 각 단어들에 대해 다음 요구사항 대로 출력하는 프로그램을 작성하시오.

>- 요구사항 1. CSS와 Javascript 내용도 모두 제거하시오 (이전 숙제에서 이미 제외하였다면 본 요구사항 무시)  
>- 요구사항 2. 순수 텍스트 안에 존재하는 단어들에 대해 string 모듈을 활용하여 모든 punctuation (구두문자)를 완벽히 제거하시오.  
>>- 예: ['world!', ':', '+hello+', '~python$$$', '=', 'world'] ---> ['world', 'hello', 'python', 'world']
>>- 모든 punctuation을 확인하는 방법은 아래와 같다.
>>> import string  
>>> print(string.punctuation)  
>- 요구사항 3. 만약 punctuation (구두문자)를 제거한 이후 공백으로 판명되는 단어들은 해당 단어 리스트에서 제거하시오.  
 예: ['컴퓨터공학부!', '---?', '%지향~', '...'] ---> ['컴퓨터공학부', '', '지향', ''] ---> ['컴퓨터공학부', '지향']  
>  &nbsp;  
>- 요구사항 4. 각 단어들의 출현빈도를 사전형태로 저장하여 출력하시오.  
예: ['world', 'hello', 'python', 'world', '컴퓨터공학부', '지향'] ---> {'world': 2, 'hello': 1, 'python': 1, '컴퓨터공학부': 1, '지향': 1}  
>  &nbsp;  
>- 요구사항 5. 다음 5개의 사이트에 대한 단어 출현빈도 결과를 출력하시오.
>>- http://cse.koreatech.ac.kr
>>- https://www.koreatech.ac.kr
>>- http://www.naver.com
>>- http://www.daum.net
>>- http://www.nytimes.com

In [29]:
import requests  # HTTP 요청 처리를 위한 requests 호출
import string    # 구두문자를 제거하기 위한 string 호출

# 엔티티 및, 이스케이프 시퀀스 제거 함수
def deleteStrinExEn(source):
    # html 엔티티값 제거
    source = source.replace("&nbsp;", "")
    source = source.replace("&lt;", "")
    source = source.replace("&gt;", "")
    source = source.replace("&amp;", "")
    source = source.replace("&apos;", "")
    source = source.replace("&quot;", "")
    source = source.replace("&lsquo;", "")
    source = source.replace("&rsquo;", "")
    source = source.replace("&ldquo;", "")
    source = source.replace("&rdquo;", "")
    source = source.replace("&deg;", "")
    source = source.replace("&middot;", "")
    source = source.replace("&times;", "")
    source = source.replace("&divide;", "")
    source = source.replace("&minus;", "")
    source = source.replace("&infin;", "")
    source = source.replace("&sim;", "")
    source = source.replace("&ne;", "")
    source = source.replace("&le;", "")
    source = source.replace("&ge;", "")
    source = source.replace("&ndash;", "")
    source = source.replace("&mdash;", "")
    source = source.replace("&bull;", "")
    source = source.replace("&hellip;", "")
    source = source.replace("&prime;", "")
    source = source.replace("&prime;", "")
    source = source.replace("&copy;", "")
    source = source.replace("&reg;", "")
    source = source.replace("&trade;", "")
    
    # html 이스케이프 시퀀스 제거
    source = source.replace("\n", "")  # source_fow 문자열 내에 '\n' 문자를 ''로 대체 (삭제)
    source = source.replace("\t", "")  # source_fow 문자열 내에 '\t' 문자를 ''로 대체 (삭제)
    source = source.replace("\r", "")  # source_fow 문자열 내에 '\r' 문자를 ''로 대체 (삭제)
    
    return source


# html 태그 제거 함수
def print_string(str1):  # str1 문자열을 인자로 받아옴
    result = []  # 반환할 문자열을 담을 리스트 result 생성
    while True:  # 모든 문자열을 추출할 때까지 반복
        i = 0  # 반복될때마다 인덱스 i = 0 으로 초기화
        # <script> 태그 제거
        if str1.find('<script>') == i:  # 문자열 중 <script>로 시작하는 인덱스가 시작과 같다면
            i = str1.find('</script>') + 9  # </script>가 닫히는 인덱스로 간 뒤 그 뒤에 인덱스 i에 저장
            str1 = str1[i:]  # i 인덱스 뒤로 슬라이싱 후 대입
            i = 0  # i = 다시 시작위치인 0으로 초기화

        # <script ~ > 태그 제거
        elif str1.find('<script ') == i:  # 문자열 중 <scirpt ~ >로 시작하는 인덱스가 시작과 같다면
            i = str1.find('</script>') + 9  # </script>가 닫히는 인덱스로 간 뒤 그 뒤에 인덱스 i에 저장
            str1 = str1[i:]  # i 인덱스 뒤로 슬라이싱 후 대입
            i = 0  # i = 다시 시작위치인 0으로 초기화

        # <SCRIPT> 태그 제거
        elif str1.find('<SCRIPT>') == i:  # 문자열 중 <SCRIPT>로 시작하는 인덱스가 시작과 같다면
            i = str1.find('</SCRIPT>') + 9  # </SCRIPT>가 닫히는 인덱스로 간 뒤 그 뒤에 인덱스 i에 저장
            str1 = str1[i:]  # i 인덱스 뒤로 슬라이싱 후 대입
            i = 0  # i = 다시 시작위치인 0으로 초기화

        # <style> 태그 제거
        elif str1.find('<style') == i:  # 문자열 중 <style>로 시작하는 인덱스가 시작과 같다면
            i = str1.find('</style>') + 8  # </style>이 닫히는 인덱스로 간 뒤 그 뒤에 인덱스 i에 저장
            str1 = str1[i:]  # i 인덱스 뒤로 슬라이싱 후 대입
            i = 0  # i = 다시 시작위치인 0으로 초기화

        # <STYLE> 태그 제거
        elif str1.find('<STYLE>') == i:  # 문자열 중 <STYLE>로 시작하는 인덱스가 시작과 같다면
            i = str1.find('</STYLE>') + 8  # </STYLE>이 닫히는 인덱스로 간 뒤 그 뒤에 인덱스 i에 저장
            str1 = str1[i:]  # i 인덱스 뒤로 슬라이싱 후 대입
            i = 0  # i = 다시 시작위치인 0으로 초기화

        # <SPAN ~ > 태그 제거
        elif str1.find('<SPAN') == i:  # 문자열 중 <SPAN ~ >로 시작하는 인덱스가 시작과 같다면
            i = str1.find('</SPAN>') + 7  # </SPAN>이 닫히는 인덱스로 간 뒤 그 뒤에 인덱스 i에 저장
            str1 = str1[i:]  # i 인덱스 뒤로 슬라이싱 후 대입
            i = 0  # i = 다시 시작위치인 0으로 초기화

        # <!-- --> 주석태그 제거
        elif str1.find('<!--') == i:  # 문자열 중 <SPAN ~ >로 시작하는 인덱스가 시작과 같다면
            i = str1.find('-->') + 3  # </SPAN>이 닫히는 인덱스로 간 뒤 그 뒤에 인덱스 i에 저장
            str1 = str1[i:]  # i 인덱스 뒤로 슬라이싱 후 대입
            i = 0  # i = 다시 시작위치인 0으로 초기화

        # 나머지 '<'' 로 시작하는 모든 태그 제거
        elif str1[i] == '<':  # 문자열 중 '<'로 시작하는 인덱스가 시작과 같다면
            while True:  # < > 내부의 중복 <>처리를 위한 반복
                i += 1  # 첫 시작 : '<' 다음, 반복시 : '>' 다음 인덱스
                if '<' in str1[i:str1.find('>') + 1]:  # <> 내부에 '<'가 있다면
                    i = str1.find('>') + 1  # 처음 '>'가 오는 인덱스 다음으로 인덱스 i 지정
                    str1 = str1[i:]  # i부터 시작하여 문자열 슬라이싱
                    i = 0  # i = 0으로 초기화
                else:  # '<' 가 중복으로 들어있지 않다면
                    i = str1.find('>') + 1  # '>'로 닫히는 인덱스로 간 뒤 그 뒤에 인덱스를 i에 저장
                    str1 = str1[i:]  # i 인덱스 뒤로 슬라이싱 후 대입
                    break  # '<' 내부에 <>가 있는지 판별하는 반복 탈출

        # 글자 또는 공백일 때
        else:  # '<'가 포함된 태그로 시작하지 않는다면
            if str1[i] == ' ':  # 공백이라면
                i += 1  # 다음 인덱스 값을 i에 저장 후
                str1 = str1[i:]  # i를 시작으로 전체 슬라이싱 후 대입
            else:  # 문자라면
                result.append(str1[i:str1.find('<')])  # 다음 태그가 시작되기 전까지 result에 문자열 대입
                i = str1.find('<')  # '<' 다음 인덱스 값을 i에 저장 후
                str1 = str1[i:]  # 다음 태그가 시작되는 인덱스부터 시작
        if len(str1) == 0:  # 모든 태그를 다 제거하였다면
            break  # 반복 종료
    return result  # 순수 문자열을 저장한 리스트 result를 반환


#숫자 및 구두문자 제거
def deletePun_Num(pathList):
    i = 0
    while i < len(pathList):
        pathList[i] = pathList[i].translate(str.maketrans('', '', string.digits))
        pathList[i] = pathList[i].translate(str.maketrans('', '', string.punctuation))
        pathList[i] = pathList[i].strip()
        i += 1
    return pathList


# 리스트 사전화
def makeDitionary(pathList):
    result = {}
    for i in pathList:
        result[i] = result.get(i, 0) + 1
    return result


# 메인
urlList = ['http://cse.koreatech.ac.kr',
           'https://www.koreatech.ac.kr',
           'http://www.naver.com',
           'http://www.daum.net',
           'http://www.nytimes.com']

for url in urlList:
    # html 문자열 저장
    req = requests.get(url)  # url의 html을 불러와서 req에 대입
    source = req.text        # 텍스트 화하여 source에 대입
    source = deleteStrinExEn(source)

    # 리스트 제작
    print("url : [ {0} ]".format(url))
    result_List = deletePun_Num(print_string(source))
    result_List = list(filter(None, result_List))  #리스트 내의 빈 공백 삭제
    result_List = ' '.join(result_List).split()
    result_Dict = makeDitionary(result_List)
    
    # print(result)   # 리스트 출력
    print('\n단어의 개수 : ', str(sum(result_Dict.values())))
    print(result_Dict)  #빈도 수 사전화
    print("\n\n")

url : [ http://cse.koreatech.ac.kr ]

단어의 개수 :  10
{'Forbidden': 2, 'You': 1, 'dont': 1, 'have': 1, 'permission': 1, 'to': 1, 'access': 1, 'this': 1, 'resource': 1}



url : [ https://www.koreatech.ac.kr ]

단어의 개수 :  609
{'한국기술교육대학교': 12, '닫기': 1, 'Link': 1, 'Service': 1, '재학생': 3, '학생이': 3, '행복한': 3, '대학': 7, '온라인교육지원': 3, '생활관': 2, '원격지원': 2, '오늘의': 2, '식단': 2, '학사종합시스템': 1, '인터넷증명발급': 3, '동아리안내': 2, '입학정보': 3, '코리아텍의': 3, '입학정보를': 3, '제공합니다': 6, '취업정보': 3, '최신': 3, '취업관련': 3, '정보를': 3, '교직원': 3, '아우누리': 1, '무선랜신청': 1, '교육장비검색': 1, '일반인': 3, '코리아텍소개': 1, 'VISION': 2, 'PLUS': 2, '언론기사': 1, '학부소개': 1, '대학원소개': 1, '캠퍼스투어': 1, '찾아오시는길': 2, '전화번호찾기': 1, '코리아텍': 1, '특장점': 1, '자세히': 1, '보기': 1, '중앙일보대학평가‘교육중심대학’': 1, '위': 4, '코리아텍은': 1, '중앙일보대학평가': 1, '년': 5, '연속으로': 1, '“교육중심대학': 1, '위”': 1, '의': 1, '성과를': 1, '이루었습니다': 1, '스킵네비게이션': 1, '본문바로가기': 1, '주메뉴바로가기': 1, '통학·': 1, '셔틀버스': 1, '안전보건경영': 2, '웹메일': 3, '행정서비스헌장': 2, '부패행위신고': 1, '홈페이지': 1, '오류신고': 1, 'ENG': 2, 'CHI': 2, '팝업': 1, '건': 1,


단어의 개수 :  821
{'Daum': 3, '주요': 2, '서비스': 5, '바로가기': 11, '오늘의': 5, '정보': 7, '미디어': 3, '실시간': 1, '이슈': 1, '멀티미디어': 2, '블로그': 4, '유익한': 2, '쇼핑': 3, '포토영상': 1, '웹': 1, '접근성': 1, '안내': 2, '자동실행': 1, '광고': 1, '설정': 1, '검색': 4, '통합': 1, '한글입력기': 1, '인기': 2, '검색어': 1, '코로나': 2, '발생현황': 1, 'BTS': 1, '온라인콘서트': 1, '벽돌판매처': 1, '맛있는녀석들': 1, '윤보미': 1, '집진기업체': 1, '사회적': 1, '거리두기': 1, '알렉볼드윈': 1, '오발사고': 1, '빅사이즈여성의류': 1, '희수': 1, '전소민': 1, '근조화환': 1, '이전': 7, '다음': 8, '다음을': 1, '시작페이지로': 1, '카페': 3, '메일': 2, '뉴스': 4, '지도': 2, '증권': 1, '카카오TV': 2, '웹툰': 2, '브런치': 3, '사전': 1, '게임': 4, '같이가치': 3, '주요서비스': 1, '더보기': 2, '전체': 5, '커뮤니케이션': 1, '티스토리': 2, '카카오스토리': 2, '연예': 2, '스포츠': 2, '금융': 1, '부동산': 2, '자동차': 2, '쇼핑생활': 1, '선물하기': 1, '쇼핑라이브': 1, '쇼핑하우': 1, '메이커스': 1, '카카오헤어샵': 1, '카카오이모티콘샵': 1, '문화엔터': 1, '멜론': 1, '영화': 2, '카카오웹툰': 1, '카카오프렌즈': 1, '소셜임팩트': 1, '프로젝트': 1, '사전지식': 1, '어학사전': 1, '번역': 1, '단어장': 1, '백과사전': 1, '배틀그라운드': 1, '패스': 1, '오브': 1, '엑자일': 1, '오딘': 1, '길찾기': 1, '로드뷰': 1, '버스정보': 1, '

>- 요구사항 1. CSS와 Javascript 내용도 모두 제거하시오 (이전 숙제에서 이미 제외하였다면 본 요구사항 무시)  
        이전 과제에서 해결
>- 요구사항 2. 순수 텍스트 안에 존재하는 단어들에 대해 string 모듈을 활용하여 모든 punctuation (구두문자)를 완벽히 제거하시오.  
       .translate(str.maketrans('', '', string.punctuation))을 활용하여 구두문자를 제거한다.   
>- 요구사항 3. 만약 punctuation (구두문자)를 제거한 이후 공백으로 판명되는 단어들은 해당 단어 리스트에서 제거하시오.  
            i = 0  
            while i < len(pathList):  
                pathList[i] = pathList[i].translate(str.maketrans('', '', string.punctuation))  
                pathList[i] = pathList[i].strip()  
                i += 1  
            을 통해 구두문자가 사라진 리스트 내 요소들의 문자열 양 옆 공백을 제거한다.
            또한,  
            result_List = list(filter(None, result_List))  
            result_List = ' '.join(result_List).split() 를 통해 리스트 내 공백을 제거한다.  
>- 요구사항 4. 각 단어들의 출현빈도를 사전형태로 저장하여 출력하시오.  
        def makeDitionary(pathList):  
            result = {}  
            for wrd in pathList:  
                result[wrd] = result.get(wrd, 0) + 1  
            return result  
            를 통해 사전화를 시킨 뒤, 중복되는 단어에 대한 value의 값을 추가해준다.  
>- 요구사항 5. 다음 5개의 사이트에 대한 단어 출현빈도 결과를 출력하시오.
        http://cse.koreatech.ac.kr
        https://www.koreatech.ac.kr
        http://www.naver.com
        http://www.daum.net
        http://www.nytimes.com

### 느낀점
 파이썬에서 사용하는 list 또는 튜플, 사전에 대하여 객체와 더불어 개념을 이해할 수 있었다.
 시퀀스 자료형을 사용하는 방법과 그에대한 객체 활용방법을 공부할 수 있어서 좋았고,
 9번 문제에서 사용한 시간복잡도 줄이는 방법은 프로그램을 얼마나 최적화하고 효과적으로 돌릴 수 있는지에 대하여 생각할 수 있게 되었다.